state/metricsmanager.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+	"gopkg.in/mgo.v2/txn"
+)
+
+var metricsManagerLogger = loggo.GetLogger("juju.state.metricsmanager")
+
+const (
+	defaultGracePeriod                      = 7 * 24 * time.Hour // 1 week in hours
+	metricsManagerConsecutiveErrorThreshold = 3
+	metricsManagerKey                       = "metricsManagerKey"
+)
+
+// MetricsManager stores data about the state of the metrics manager
+type MetricsManager struct {
+	st  *State
+	doc metricsManagerDoc
+}
+
+type metricsManagerDoc struct {
+	DocID              string        `bson:"_id"`
+	LastSuccessfulSend time.Time     `bson:"lastsuccessfulsend"`
+	ConsecutiveErrors  int           `bson:"consecutiveerrors"`
+	GracePeriod        time.Duration `bson:"graceperiod"`
+}
+
+// DocID returns the Document id of the MetricsManager.
+func (m *MetricsManager) DocID() string {
+	return m.doc.DocID
+}
+
+// LastSuccessfulSend returns the time of the last successful send.
+func (m *MetricsManager) LastSuccessfulSend() time.Time {
+	return m.doc.LastSuccessfulSend
+}
+
+// ConsecutiveErrors returns the number of consecutive failures.
+func (m *MetricsManager) ConsecutiveErrors() int {
+	return m.doc.ConsecutiveErrors
+}
+
+// GracePeriod returns the current grace period.
+func (m *MetricsManager) GracePeriod() time.Duration {
+	return m.doc.GracePeriod
+}
+
+// MetricsManager returns an existing metricsmanager, or a new one if non exists.
+func (st *State) MetricsManager() (*MetricsManager, error) {
+	mm, err := st.getMetricsManager()
+	if errors.IsNotFound(err) {
```

Comment:
I think you should fold `newMetricsManager` in here, and change the buildTxn around a bit to do:

loop:
    mm, err := st.getMetricsManager()
    if !errors.IsNotFound(err) {
        return mm, err
    }
    return op that creates if not exists